lib/async_hooks.js

```diff
@@ -209,11 +211,102 @@ class AsyncResource {
   }
 }
 
+const storageList = [];
+const storageHook = createHook({
+  init(asyncId, type, triggerAsyncId, resource) {
+    const currentResource = executionAsyncResource();
+    // Value of currentResource is always a non null object
+    for (let i = 0; i < storageList.length; ++i) {
+      storageList[i]._propagate(resource, currentResource);
+    }
+  }
+});
+
+class AsyncLocalStorage {
+  constructor() {
+    this.kResourceStore = Symbol('kResourceStore');
+    this.enabled = false;
+  }
+
+  disable() {
+    if (this.enabled) {
+      this.enabled = false;
+      // If this.enabled, the instance must be in storageList
+      storageList.splice(storageList.indexOf(this), 1);
+      if (storageList.length === 0) {
+        storageHook.disable();
+      }
+    }
+  }
+
+  // Propagate the context from a parent resource to a child one
+  _propagate(resource, triggerResource) {
+    const store = triggerResource[this.kResourceStore];
+    if (this.enabled) {
+      resource[this.kResourceStore] = store;
+    }
+  }
+
+  _enter() {
```

Comment:
Might be nice to change `_enter()` to `_enter(store = new Map())` and then make `run*WithStore(store, callback, ...args)` methods to allow using something _other_ than the `Map` instance as the "store" object. Handy for APMs that already have their own existing request data class they need to store, that way you don't need the extra `Map` allocation and lookup in the way when you only have the one thing to store.

Reply:
It could be a nice enhancement. As an option, both `.run*()` methods could be changed to `.run*(callback, store)` with optional second argument (`store`) which is used as the store when provided. But in that case, we'll have to sacrifice `...args` feature of those methods. On the other hand, Cartesian product of `.run*()` methods (i.e. 4 methods instead of 2) is not ideal option as well.

Reply:
What about the change suggested in https://github.com/nodejs/node/pull/26540#issuecomment-585421990 ?

Reply:
I'd rather go with public `.enter([store])`/`.exit()` methods than `.setStore(store)`.

To add to already mentioned options, I can also think of another way to provide custom objects to be used as the store. Namely, a new `.withStore(store)` method that works as an override to be used only for the next `.run*()` invocation:
```js
asyncLocalStorage
  .withStore({ foo: 'bar' })
  .run(() => { /* ... */ }); // { foo: 'bar' } used here

asyncLocalStorage.run(() => { /* ... */ }); // a new Map used here
```

Reply:
@Flarna wdyt here? I don't have strong opinions

Reply:
Adding `store` as first argument to the `run*` calls sounds fine as there is still the ability to just wrap existing callbacks without the need to modify their signature.
Using a dedicated `.withStore()` function seems to be more error prone as it is not that clear if this store is valid just for the next `.run` call or till `setStore()` is called again,...?

Reply:
Thanks a lot for the great inputs from all of you! At the point, we agree on the goal but are not sure about the method. And this should only block us if it's something that can't be iterated upon.

I think it's not the right time to go with public `enter/exit`, we have iterated too much and the current state of the PR is already a compromise between a lot of PoW and that is always something we can add later.

I don't want to change this PR too much as multiple people have already approved it.

TBH, as of today, I forsee non-breaking solutions for iterating over this (wether it's `run*([store,] cb, ...args)`, `.withStore()`, `setStore` or even something else). At this point I hope to have this merged next week (after https://github.com/vdeturckheim/node/pull/4 is merged and CI fixed ofc).

Reply:
For what it's worth, adding `cls.runWithStore(store, callback, ...args)` make it so `cls.run(callback, ...args)` can be implemented simply as:

```js
run(callback, ...args) {
  return this.runWithStore(new Map(), callback, ...args)
}
```

Doesn't do anything about the cognitive burden of more methods for the end-user to think about, but it _does_ mean there's not much maintenance burden to worry about.

Reply:
By the way, I'm not intending this comment to block landing. I just wanted to mention the idea here so we think about it and have a path forward to add a feature like this after this PR lands. I'm happy to implement it myself too, if you don't want to spend the time on it. :)

Reply:
I agree that `.runWithStore*()` methods can be added in a subsequent PR. But those methods will increase amounts of things that need to be documented and tested, so it's not only about implementation (which is trivial). Moreover, users might be confused by so many similar methods. So, before we go that road I'd like to propose another approach.

Why don't we just always ask for a store in `.run*()` methods instead of creating a `Map` implicitly? Signature of those methods can be simply changed to `.run*(store, callback[, ..args])`. Users who want to use a `Map` will simply write `asyncLocalStorage.run(new Map(), () => {/*...*/})`. IMO most users are going to use a custom object with some data (e.g. request id) when calling `.run*()`, so they don't really need an extra `Map`.

Reply:
@puzpuzpuz as mentionned in the few previous messages, we should probably see that in an upcoming PR. I really like the `run*([store,] cb, ...args)` approach but I'd like to avoid changing this PR too much after reviewer have already validated it.

Also, this PR is the 2nd most discussed PR in the project so far ^^

Reply:
@vdeturckheim I understand your point about existing code reviews.

I proposed that change (`.run*(store, callback[, ..args])`) because it's an API compatibility breaking change, plus it's quite trivial and doesn't assume significant changes in the implementation. Do you think it's OK to change that in a subsequent PR once this one is merged?

Reply:
I am not sure here. I guess I think `.run*(store, callback[, ..args])` can work as `.run*([store,] callback[, ..args])`, would that one be acceptable for you?

Indeed, the process here is not breaking change friendly and I am not sure what are the rules regarding "breaking changes in unpublished APIs".

Reply:
> I am not sure here. I guess I think `.run*(store, callback[, ..args])` can work as `.run*([store,] callback[, ..args])`, would that one be acceptable for you? Also `run*WithStore` makes sense to me.

I don't see how `.run*([store,] callback[, ..args])` can be implemented, as current `.run*(callback[, ...args])` methods can have arbitrary number of arguments plus `store` itself may be a function. Do you see a way to implement it?

Reply:
Yes, I would just have a check on `typeof arg0 === 'function'`. If this is true, the first argument is the callback, otherwise it is the store.

Reply:
> Yes, I would just have a check on `typeof arg0 === 'function'`. If this is true, the first argument is the callback, otherwise it is the store.

I'm afraid that this approach is error prone, as some users may want to use a function as the store. I'd rather go with `.run*WithStore()` than `.run*([store,] callback[, ..args])`, if I have to pick a backwards-compatible change. Of course, both options aren't ideal.

Reply:
I would have the opinionated standpoint that using a function as a store is a bad practice ^^

Reply:
IMO core API should not make such assumptions and implement functions with ambivalent optional arguments. In any case, we can re-iterate over the approach once this PR is merged.

Reply:
Let's iterate :D